### PR TITLE
Sike June updates

### LINKED ISSUE #1003: SIKE fails to compile with gcc 11.1.0
SIKE on the main branch fails to compile on Arch Linux with gcc 11.1.0. Clang 11.1.0 (apparently GCC has caught up with Clang in version numbers) is fine.

https://gist.github.com/thomwiggers/1914f30498bea7d7ee9a1bc9b2d66db4

The problem appears to lie in `-Warray-parameter=2`, which is included in `-Wall`.

>  At level 2 the warning also triggers for redeclarations involving any other inconsistency in array or pointer argument forms denoting array sizes.  Pointers and arrays of unspecified bound are considered equivalent and do not
           trigger a warning.
>
>       void g (int*);
>       void g (int[]);     // no warning
>       void g (int[8]);    // warning (inconsistent array bound)

```
In file included from ../src/kem/sike/external/P503/P503.c:164:
../src/kem/sike/external/P503/../fpx.c:90:35: error: argument 1 of type ‘const digit_t[8]’ {aka ‘const long unsigned int[8]’} with mismatched bound [-Werror=array-parameter=]
   90 | __inline void fpcopy(const felm_t a, felm_t c)
      |                      ~~~~~~~~~~~~~^
In file included from ../src/kem/sike/external/P503/P503.c:10:
../src/kem/sike/external/P503/P503_internal.h:148:38: note: previously declared as ‘const digit_t *’ {aka ‘const long unsigned int *’}
  148 | static void fpcopy503(const digit_t *a, digit_t *c);
      |                       ~~~~~~~~~~~~~~~^
```

It also fails on `-Warray-bounds`

```
In file included from ../src/kem/sike/external/P751/P751_compressed.c:435:
../src/kem/sike/external/P751/../compression/dlog.c: In function ‘Traverse_w_div_e_torus.constprop’:
../src/kem/sike/external/P751/../compression/dlog.c:166:13: error: array subscript 3 is above array bounds of ‘digit_t[3][2][12]’ {aka ‘long unsigned int[3][2][12]’} [-Werror=array-bounds]
  166 |         d = mod(ord2w_dlog(&H[W_2_1][0], logT, Texp), (1 << W_2_1));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/kem/sike/external/P751/../compression/dlog.c:141:13: note: while referencing ‘H’
  141 |     f2elm_t H[W_2_1] = {0}; // Size of H should be max of {W_2_1, W_2 - W_2_1}
      |             ^
```

And on `-Wstringop-overflow`:
```
In function ‘CompleteMPoint’,
    inlined from ‘BuildOrdinary2nBasis_dual’ at ../src/kem/sike/external/P751/../compression/torsion_basis.c:440:5,
    inlined from ‘EphemeralKeyGeneration_B_extended’ at ../src/kem/sike/external/P751/../compression/sidh_compressed.c:630:5:
../src/kem/sike/external/P751/P751_compressed.c:400:21: error: ‘fp2mul751_mont’ accessing 192 bytes in a region of size 96 [-Werror=stringop-overflow=]
  400 | #define fp2mul_mont fp2mul751_mont
../src/kem/sike/external/P751/../ec_isogeny.c:434:9: note: in expansion of macro ‘fp2mul_mont’
  434 |         fp2mul_mont(P->X, P->Z, xz);       // xz = x*z;
      |         ^~~~~~~~~~~
../src/kem/sike/external/P751/../ec_isogeny.c: In function ‘EphemeralKeyGeneration_B_extended’:
../src/kem/sike/external/P751/P751_compressed.c:400:21: note: referencing argument 2 of type ‘const digit_t (*)[12]’ {aka ‘long unsigned int (*)[12]’}
  400 | #define fp2mul_mont fp2mul751_mont
../src/kem/sike/external/P751/../ec_isogeny.c:434:9: note: in expansion of macro ‘fp2mul_mont’
  434 |         fp2mul_mont(P->X, P->Z, xz);       // xz = x*z;
      |         ^~~~~~~~~~~
../src/kem/sike/external/P751/P751_compressed.c:400:21: note: in a call to function ‘fp2mul751_mont’
  400 | #define fp2mul_mont fp2mul751_mont
      |                     ^~~~~~~~~~~~~~
../src/kem/sike/external/P751/../fpx.c:320:13: note: in expansion of macro ‘fp2mul_mont’
  320 | static void fp2mul_mont(const f2elm_t a, const f2elm_t b, f2elm_t c)
      |             ^~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by christianpaquin**
Integrated SIKE's latest commits, up to [28b4b5](https://github.com/microsoft/PQCrypto-SIDH/commit/28b4b5d0a7926e0e7eb4f9c03f75887236e1cebd) ([diff](https://github.com/microsoft/PQCrypto-SIDH/compare/10b0f4...28b4b5)).

Fixes #1003.

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the the list of algorithms available -- either adding, removing, or renaming?  (If so, PRs in OQS-OpenSSL, OQS-BoringSSL, and OQS-OpenSSH will also be required by the time this is merged.)

### CODE REVIEW

**Comment by thomwiggers** [comment]
This does seem to solve the SIKE issues with gcc 11.

**Comment by thomwiggers** [comment]
`OQS_DIST_BUILD=Yes` breaks though, but I assume that's why CI is not happy.

File: src/kem/sike/external/fpx.c [hunk 1]
```diff
@@ -280,15 +280,15 @@ __inline unsigned int mp_sub(const digit_t* a, const digit_t* b, digit_t* c, con
...
+#if (OS_TARGET == OS_WIN) || defined(GENERIC_IMPLEMENTATION) || (TARGET == TARGET_ARM)
     felm_t t1;
 
     digit_t mask = 0 - (digit_t)mp_sub(a, b, c, 2*NWORDS_FIELD);
     for (int i = 0; i < NWORDS_FIELD; i++)
         t1[i] = ((digit_t*)PRIME)[i] & mask;
     mp_addfast((digit_t*)&c[NWORDS_FIELD], t1, (digit_t*)&c[NWORDS_FIELD]);
 
-#else
+#elif (OS_TARGET == OS_NIX)
```
Comment (jschanck) [inline src/kem/sike/external/fpx.c:291]: The P434 and P610 parameter sets only have x86-64 assembly versions of mp_subaddx2 if the CPU supports BMI2 instructions, so this check is not sufficient. There are similar problems for the conditions that wrap calls to mp_dblsub610x2, mp_add610, and the 434 versions of these.

Comment (christianpaquin) [reply src/kem/sike/external/fpx.c:291]: Ah yes, my last update overwrote correct checks for this; I'll revert them back.

**Review by dstebila — APPROVED** [review]

**Pull request merged**